crates/utils/src/interval_map/mod.rs

```diff
@@ -0,0 +1,902 @@
+        if x.right(NodeRef::is_sentinel) {
+            return;
+        }
+        let y = x.right_owned();
+        x.set_right(y.left_owned());
+        if !y.left(NodeRef::is_sentinel) {
+            y.left(|l| l.set_parent(x.clone_rc()));
+        }
+
+        self.replace_parent(x, &y);
+        y.set_left(x.clone_rc());
+
+        Self::rotate_update_max(x, &y);
+    }
+
+    /// Binary tree right rotate.
+    fn right_rotate(&mut self, x: &NodeRef<T, V>) {
+        if x.left(NodeRef::is_sentinel) {
+            return;
+        }
+        let y = x.left_owned();
+        x.set_left(y.right_owned());
+        if !y.right(NodeRef::is_sentinel) {
+            y.right(|r| r.set_parent(x.clone_rc()));
+        }
+
+        self.replace_parent(x, &y);
+        y.set_right(x.clone_rc());
+
+        Self::rotate_update_max(x, &y);
+    }
+
+    /// Replaces parent during a rotation.
+    fn replace_parent(&mut self, x: &NodeRef<T, V>, y: &NodeRef<T, V>) {
+        y.set_parent(x.parent_owned());
+        if x.parent(NodeRef::is_sentinel) {
+            self.root = y.clone_rc();
+        } else if x.parent(|p| p.left(|l| l.eq(x))) {
+            x.parent(|p| p.set_left(y.clone_rc()));
+        } else {
+            x.parent(|p| p.set_right(y.clone_rc()));
+        }
+        x.set_parent(y.clone_rc());
+    }
+
+    /// Updates the max value after a rotation.
+    fn rotate_update_max(x: &NodeRef<T, V>, y: &NodeRef<T, V>) {
+        y.set_max(x.max_owned());
+        let mut max = x.interval(|i| i.high.clone());
+        if !x.left(NodeRef::is_sentinel) {
+            max = max.max(x.left(NodeRef::max_owned));
+        }
+        if !x.right(NodeRef::is_sentinel) {
+            max = max.max(x.right(NodeRef::max_owned));
+        }
+        x.set_max(max);
+    }
+
+    /// Updates the max value towards the root
+    fn update_max_bottom_up(x: &NodeRef<T, V>) {
+        let mut p = x.clone_rc();
+        while !p.is_sentinel() {
+            p.set_max(p.interval(|i| i.high.clone()));
+            Self::max_from(&p, &p.left_owned());
+            Self::max_from(&p, &p.right_owned());
+            p = p.parent_owned();
+        }
+    }
+
+    /// Updates a nodes value from a child node.
+    fn max_from(x: &NodeRef<T, V>, c: &NodeRef<T, V>) {
+        if !c.is_sentinel() && x.max(|xm| c.max(|cm| xm < cm)) {
+            x.set_max(c.max_owned());
+        }
+    }
+
+    /// Finds the node with the minimum interval.
+    fn tree_minimum(mut x: NodeRef<T, V>) -> NodeRef<T, V> {
+        loop {
+            let left = x.left_owned();
+            if left.is_sentinel() {
+                return x;
+            }
+            x = left;
+        }
+    }
+
+    /// Replaces one subtree as a child of its parent with another subtree.
+    fn transplant(&mut self, u: &NodeRef<T, V>, v: &NodeRef<T, V>) {
+        if u.parent(NodeRef::is_sentinel) {
+            self.root = v.clone_rc();
+        } else {
+            if u.parent(|p| p.left(|l| l == u)) {
```

Comment:
Please replace this condition with `is_left_child`.

Reply:
done